### ISSUE TITLE
Redis backend connecting to localhost instead of provided host

### ISSUE THREAD

**Issue by pb-jeff-oneill** [reclosedev/requests-cache#611]
I've been using requests cache for awhile and have been loving it.  Big thanks to the authors!

I've got a weird issue that may be a bug.

Running this code:
```
redis_host='1.2.3.4'
redis_port='6379'
redis_client = redis.StrictRedis(host=redis_host, port=redis_port)
backend = RedisCache(host=redis_host, port=redis_port)
cached_requests = CachedSession('http_cache', backend=backend, expire_after=24*60*60)
```
gives the traceback below.  It seems to be trying to connect to localhost instead of the provided redis host.

I'm on:
- Python 3.7
- requests-cache==0.9.3
- redis==4.2.2
- Also running Google App Engine in case it matters.  Redis is otherwise working properly there.

Any idea why `CachedSession` is using localhost instead of the provided host?

> Traceback (most recent call last):
>   File "/.../python3.7/site-packages/redis/connection.py", line 615, in connect
>     lambda: self._connect(), lambda error: self.disconnect(error)
>   File "/.../python3.7/site-packages/redis/retry.py", line 45, in call_with_retry
>     return do()
>   File "/.../python3.7/site-packages/redis/connection.py", line 615, in <lambda>
>     lambda: self._connect(), lambda error: self.disconnect(error)
>   File "/.../python3.7/site-packages/redis/connection.py", line 680, in _connect
>     raise err
>   File "/.../python3.7/site-packages/redis/connection.py", line 668, in _connect
>     sock.connect(socket_address)
> ConnectionRefusedError: [Errno 61] Connection refused
> 
> During handling of the above exception, another exception occurred:
> 
> Traceback (most recent call last):
>   File "<stdin>", line 1, in <module>
>   File "/.../python3.7/site-packages/requests_cache/session.py", line 60, in __init__
>     self.cache = init_backend(cache_name, backend, **kwargs)
>   File "/.../python3.7/site-packages/requests_cache/backends/__init__.py", line 76, in init_backend
>     logger.debug(f'Initializing backend: {backend} {cache_name}')
>   File "/.../python3.7/site-packages/requests_cache/backends/base.py", line 235, in __str__
>     return f'Total rows: {len(self.responses)} responses, {len(self.redirects)} redirects'
>   File "/.../python3.7/site-packages/requests_cache/backends/redis.py", line 108, in __len__
>     return len(list(self.keys()))
>   File "/.../python3.7/site-packages/requests_cache/backends/redis.py", line 121, in keys
>     for key in self.connection.keys(f'{self.namespace}:*')
>   File "/.../python3.7/site-packages/redis/commands/core.py", line 1745, in keys
>     return self.execute_command("KEYS", pattern, **kwargs)
>   File "/.../python3.7/site-packages/redis/client.py", line 1215, in execute_command
>     conn = self.connection or pool.get_connection(command_name, **options)
>   File "/.../python3.7/site-packages/redis/connection.py", line 1386, in get_connection
>     connection.connect()
>   File "/.../python3.7/site-packages/redis/connection.py", line 620, in connect
>     raise ConnectionError(self._error_message(e))
> redis.exceptions.ConnectionError: Error 61 connecting to **localhost:6379**. Connection refused.

**Comment by pb-jeff-oneill** [comment]
After digging into the code, this is a documentation bug.  This is the fix:

`backend = RedisCache(connection=redis_client)`

The docs [here](https://requests-cache.readthedocs.io/en/stable/modules/requests_cache.backends.redis.html#connection-options) use the older interface.

**Comment by JWCook** [comment]
@pb-jeff-oneill Your original example is intended to work, but appears to have recently broken.

`RedisCache` will use inspection to determine valid keyword args that can be passed along to `Redis`. Strangely, that doesn't work in redis-py 4.2 and python <=3.8, as the signature of `Redis.__init__()` no longer lists its keyword args. I did add a version restriction in pyproject.toml, but that hasn't been released.

The best option, as you found, is to pass a `Redis` object.

**Comment by JWCook** [comment]
I think I will keep this open as a bug for now, since this is likely to be confusing to other users.

**Comment by JWCook** [comment]
This is actually a fairly interesting one.

# Short version
Just need to inspect the `Redis.__init__` method instead of the `Redis` class.

# Long version
The behavior of `inspect.signature` for classes isn't well documented, but in most cases it will get the signature of the class's init method. This is an odd case where that doesn't happen. Specifically on redis-py 4.2 and python 3.7 and 3.8, it looks like it will get the signature of `__new__` instead, even though one isn't explicitly defined on the `Redis` class or its parents.

So it appears that:
* The behavior of `inspect.signature` for classes was improved in python 3.9:
  - https://github.com/python/cpython/issues/85074
  - https://github.com/python/cpython/pull/27209
* The inheritance order of `redis.client.Redis` [changed in redis-py 4.2](https://github.com/redis/redis-py/compare/v4.1.4...v4.2.0#diff-3161b005adfa1f5266a9736bffb5ede9d14e2ab40937cb241bc16ca5647df79d) (now inherits from an abstract class), although I still don't see how that affects its signature.
* This also seems to be the case with `pymongo.mongo_client.MongoClient` in pymongo 4.1.

# Examples
redis-py 4.2.0 on python 3.8:
```python
from inspect import signature

signature(Redis.__new__).parameters
mappingproxy({'cls': <Parameter "cls">, 'args': <Parameter "*args">, 'kwds': <Parameter "**kwds">})

signature(Redis).parameters
mappingproxy({'args': <Parameter "*args">, 'kwds': <Parameter "**kwds">})
```

redis-py 4.1.4 with any python version, **OR** python 3.9 with any redis-py version:
```python
from inspect import signature

signature(Redis.__new__).parameters
mappingproxy({'args': <Parameter "*args">, 'kwargs': <Parameter "**kwargs">})

signature(Redis).parameters
mappingproxy({
    'host': <Parameter "host='localhost'">,
    'port': <Parameter "port=6379">,
    'db': <Parameter "db=0">,
    'password': <Parameter "password=None">,
    'socket_timeout': <Parameter "socket_timeout=None">,
    'socket_connect_timeout': <Parameter "socket_connect_timeout=None">,
    'socket_keepalive': <Parameter "socket_keepalive=None">,
    'socket_keepalive_options': <Parameter "socket_keepalive_options=None">,
    'connection_pool': <Parameter "connection_pool=None">,
    'unix_socket_path': <Parameter "unix_socket_path=None">,
    'encoding': <Parameter "encoding='utf-8'">,
    'encoding_errors': <Parameter "encoding_errors='strict'">,
    'charset': <Parameter "charset=None">,
    'errors': <Parameter "errors=None">,
    'decode_responses': <Parameter "decode_responses=False">,
    'retry_on_timeout': <Parameter "retry_on_timeout=False">,
    'retry_on_error': <Parameter "retry_on_error=[]">,
    'ssl': <Parameter "ssl=False">,
    'ssl_keyfile': <Parameter "ssl_keyfile=None">,
    'ssl_certfile': <Parameter "ssl_certfile=None">,
    'ssl_cert_reqs': <Parameter "ssl_cert_reqs='required'">,
    'ssl_ca_certs': <Parameter "ssl_ca_certs=None">,
    'ssl_ca_path': <Parameter "ssl_ca_path=None">,
    'ssl_ca_data': <Parameter "ssl_ca_data=None">,
    'ssl_check_hostname': <Parameter "ssl_check_hostname=False">,
    'ssl_password': <Parameter "ssl_password=None">,
    'ssl_validate_ocsp': <Parameter "ssl_validate_ocsp=False">,
    'ssl_validate_ocsp_stapled': <Parameter "ssl_validate_ocsp_stapled=False">,
    'ssl_ocsp_context': <Parameter "ssl_ocsp_context=None">,
    'ssl_ocsp_expected_cert': <Parameter "ssl_ocsp_expected_cert=None">,
    'max_connections': <Parameter "max_connections=None">,
    'single_connection_client': <Parameter "single_connection_client=False">,
    'health_check_interval': <Parameter "health_check_interval=0">,
    'client_name': <Parameter "client_name=None">,
    'username': <Parameter "username=None">,
    'retry': <Parameter "retry=None">,
    'redis_connect_func': <Parameter "redis_connect_func=None">
})
```

**Comment by JWCook** [comment]
This has now been fixed in v0.9.4.
